src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java

```diff
@@ -54,35 +73,132 @@ public LostItemFeeChargingService(Clients clients) {
       loan, request, staffUserId);
+    final AccountCancelReason reason = AccountCancelReason.CANCELLED_ITEM_DECLARED_LOST;
+
+    log.info("service point data: " + request.getServicePointId());
 
     return lostItemPolicyRepository.getLostItemPolicyById(loan.getLostItemPolicyId())
       .thenApply(result -> result.map(referenceDataContext::withLostItemPolicy))
       .thenCompose(refDataResult -> refDataResult.after(referenceData -> {
         if (shouldCloseLoan(referenceData.lostItemPolicy)) {
           log.debug("Loan [{}] can be closed because no fee will be charged", loan.getId());
-          return closeLoanAndUpdateInStorage(loan)
-            .thenCompose(r -> r.after(eventPublisher::publishClosedLoanEvent))
-            .thenApply(r -> r.map(v -> loan));
+          return closeLoanAsLostAndPaidAndUpdateInStorage(loan)
+          .thenCompose(r -> r.after(eventPublisher::publishClosedLoanEvent))
+          .thenApply(r -> r.map(v -> loan));
         }
 
-        return fetchFeeFineOwner(referenceData)
-          .thenApply(this::refuseWhenFeeFineOwnerIsNotFound)
-          .thenComposeAsync(this::fetchFeeFineTypes)
-          .thenApply(this::buildAccountsAndActions)
-          .thenCompose(r -> r.after(feeFineFacade::createAccounts))
-          .thenApply(r -> r.map(notUsed -> loan));
+        log.info("Checking for existing lost item fees for loan [{}]", loan.getId());
+        return accountRepository.findAccountsForLoan(loan)
+          .thenCompose(result -> {
+            this.loanWithAccountData = result.value();
+            if (!hasLostItemFees(loanWithAccountData)) {
+              log.info("No existing lost item fees found, applying lost item fees to loan [{}]", loan.getId());
+              return applyFees(referenceData, loanWithAccountData);
+            }
+            log.info("Existing lost item fees found for loan [{}], trying to clear", loan.getId());
+            final LostItemFeeRefundContext refundContext = new LostItemFeeRefundContext(
+              loan.getItem().getStatus(),
+              loan.getItem().getItemId(),
+              userId,
+              servicePointId,
+              loanWithAccountData,
+              reason
+            );
+            // refund service will refund and close any partially paid or transferred accounts.
+            // accounts that have not been processed (paid or transferred) require a separate
+            // process to close
+            return refundService.refundAccounts(refundContext)
+              .thenCompose(r -> {return cancelAccounts(loanWithAccountData);})
+              .thenCompose(res -> {return accountRepository.findAccountsForLoan(loan);})
+              .thenCompose(loanResult -> {
+                if (hasLostItemFees(loanResult.value())) {
+                  log.error("Cancel/refund fees unsuccessful, aborting charging new fees");
+                  return CompletableFuture.completedFuture(succeeded(loan));
```

Comment:
Because of the refund process code.  The refund process code fails in two situations: 1.  it could not refund the fees, 2.  It could not find any partially-paid accounts to refund.  The second case is not an "error" in the context of this process, but it does fail the result.  so rather than write a lot of complicated code to trap that specific situation and pass it, I decided to just write code that checked after the entire refund/cancellation process and see if there were still remaining accounts.  Granted, for cancellation, it's not really necessary, because that code works very differently from refunds-a failure there is actually a failure to cancel accounts.  But if i'm going to do an additional check, it just made more sense to me to position it at the end of the entire process.

Reply:
> Because of the refund process code. The refund process code fails in two situations: 1. it could not refund the fees, 2. It could not find any partially-paid/transferred accounts to refund (usually because there are none). The second case is not an "error" in the context of this process, but it does fail the result. so rather than write a lot of complicated code to trap that specific situation and pass it, I decided to just write code that checked after the entire refund/cancellation process and see if there were still remaining accounts. Granted, for cancellation, it's not really necessary, because that code works very differently from refunds-a failure there is actually a failure to cancel accounts. But if i'm going to do an additional check, it just made more sense to me to position it at the end of the entire process.

I think I'm a bit confused.

Let me try to understand the flow of this code: 
# Refund outstanding accounts related to this loan
# Ignore the outcome of that process
# Cancel the accounts related to this loan
# Ignore the outcome of that process
# Fetch the accounts related to this loan
# If there are any lost item fees, log an error, however consider the overall process a success
# otherwise issues some new fees / fines

Is that anywhere close?

This feels like it is rather complicated. Have you checked  with Holly that this is the expected behaviour?

Reply:
That looks accurate.  Yes, it absolutely is complicated.  I'm unclear on what you mean by "the expected behavior."  the story specifies attempting to refund fees first (being done) and then to attempt to cancel accounts (done).  What's not specified in the story is how to handle situations where, for whatever reason, the accounts can't be cleared.  I did ask about this but I didn't really get a clear answer.  However, as a matter of practicality, my options are limited.  At this point in the process, it looks like the loan status has already been changed to declared lost, so I can't fail the entire declare lost process (can I?).  So my options there are limited to leaving the unclearable fees and charge new ones or not.

Reply:
> However, as a matter of practicality, my options are limited. At this point in the process, it looks like the loan status has already been changed to declared lost, so I can't fail the entire declare lost process (can I?)

I agree, I think that, in reality, these processes can partly succeed / fail. This is a complicated area that FOLIO hasn't really addressed much (there are spikes for me to somehow answer these kinds of topics for processes like check out and check in).

I suggest writing up the decisions that you've made (feel free to take what I wrote above if it's correct) and publish it on the issue and ask Holly if that is what she would expect to happen e.g. that the process should continue if the refunding step fails etc.

Does that make sense?

Reply:
Yes, I can absolutely do that.